app/sdil/controllers/ReturnsController.scala

```diff
@@ -197,21 +216,82 @@ class ReturnsController (
   } yield sdilReturn
 
+  implicit class LongComparison(x :(Long,Long)) {
+    def > (y:Long): Boolean = {
+        x._1 + x._2 > y
+    }
+  }
+
+  implicit class SmallProducerDetails(smallProducers: List[SmallProducer]) {
+    def total: (Long, Long) = smallProducers.map(x => x.litreage).combineAll
+  }
+
+  private def askNewWarehouses()(implicit hc: HeaderCarrier): WebMonad[List[Site]] = for {
+    addWarehouses  <- ask(bool, "ask-secondary-warehouses-in-return")(implicitly, implicitly, extraMessages)
+    firstWarehouse <- ask(warehouseSiteMapping,"first-warehouse")(warehouseSiteForm, implicitly, ExtraMessages()) when
+      addWarehouses
+    warehouses     <- askWarehouses(firstWarehouse.fold(List.empty[Site])(x => List(x))) emptyUnless
+      addWarehouses
+  } yield warehouses
+
+  private def askNewPackingSites(subscription: Subscription)(implicit hc: HeaderCarrier): WebMonad[List[Site]] = {
+    val extraMessages  = ExtraMessages(
+      messages = Map(
+        "pack-at-business-address-in-return.lead" -> s"${Address.fromUkAddress(subscription.address).nonEmptyLines.mkString("<br/>")}")
+    )
+    for {
+      usePPOBAddress   <- ask(bool, "pack-at-business-address-in-return")(implicitly, implicitly, extraMessages)
+      packingSites     = if (usePPOBAddress) {
+        List(Site.fromAddress(Address.fromUkAddress(subscription.address)))
+      } else {
+        List.empty[Site]
+      }
+      firstPackingSite <- ask(packagingSiteMapping,"first-production-site")(packagingSiteForm, implicitly, ExtraMessages()) when
+        packingSites.isEmpty
+      packingSites     <- askPackSites(packingSites ++ firstPackingSite.fold(List.empty[Site])(x => List(x)))
+    } yield packingSites
+  }
+
   private def program(period: ReturnPeriod, subscription: Subscription, sdilRef: String)
                      (implicit hc: HeaderCarrier): WebMonad[Result] = for {
     sdilReturn     <- askReturn(subscription, sdilRef)
+    // check if they need to vary
+    isNewImporter   = (sdilReturn.importLarge |+| sdilReturn.importSmall) > 0 && !subscription.activity.importer
+    isNewPacker     = (sdilReturn.packLarge |+| sdilReturn.packSmall.total) > 0 && !subscription.activity.contractPacker
+    inner           = uniform.fragments.return_variation_continue(isNewImporter, isNewPacker)
+    _               <- tell("return-change-registration", inner) when isNewImporter || isNewPacker
+    newPackingSites <- askNewPackingSites(subscription) when isNewPacker && subscription.productionSites.isEmpty
+    newWarehouses   <- askNewWarehouses when isNewImporter && subscription.warehouseSites.isEmpty
+
+    variation     = ReturnsVariation(
+      orgName = subscription.orgName,
+      ppobAddress = subscription.address,
+      importer = (isNewImporter, (sdilReturn.importLarge |+| sdilReturn.importSmall).combineN(4)),
+      packer = (isNewPacker, (sdilReturn.packLarge |+| sdilReturn.packSmall.total).combineN(4)),
```

Comment:
maybe consider having `totalPacked` and `totalImported` methods on sdilReturn if you're using them a lot?

Reply:
Hmm, not sure, should the variation data not be distinct from the return data, even if it is derived from it?